Mathlib/Tactic/Linter/UnusedInstancesInType.lean

```diff
@@ -0,0 +1,233 @@
+  let unusedInstanceBinders := unusedInstanceBinders.map toMessageData
+  m!"`{.ofConstName declName}` has the \
+  {if unusedInstanceBinders.size = 1 then "hypothesis" else "hypotheses"}:\
+  {(unusedInstanceBinders.map (m!"\n  • {·}") |>.foldl (init := .nil) .compose)}\nwhich \
+  {if unusedInstanceBinders.size = 1 then "is" else "are"} not used in the remainder of the type."
+
+/--
+Gathers instance hypotheses in the type of `decl` that are unused in the remainder of the type and
+whose types satisfy `p`. (Does not consider the body of the declaration.) Collects them into an
+`Array Parameter`, and if (and only if) this array is nonempty, feeds it to `logOnUnused`.
+
+Since `logOnUnused` will only be run if its argument is nonempty, it is allowed to be expensive.
+
+Note that `p` is non-monadic, and may encounter loose bvars in its argument. This is a performance
+optimization. However, the `Parameter`s are created in a telescope, and their fields will *not*
+have loose bound variables.
+-/
+def _root_.Lean.ConstantVal.onUnusedInstancesWhere (decl : ConstantVal)
+    (p : Expr → Bool) (logOnUnused : Array Parameter → TermElabM Unit) : CommandElabM Unit := do
+  let unusedInstances := decl.type.getUnusedForallInstanceBinderIdxsWhere p
+  if let some maxIdx := unusedInstances.back? then liftTermElabM do
+    unless decl.type.hasSorry do -- only check for `sorry` in the "expensive" case
+      forallBoundedTelescope decl.type (some <| maxIdx + 1)
+        (cleanupAnnotations := true) fun fvars _ => do
+          let unusedInstances : Array Parameter ← unusedInstances.mapM fun idx =>
+            return {
+                fvar? := fvars[idx]?
+                type? := ← fvars[idx]?.mapM (inferType ·)
+                idx
+              }
+          logOnUnused unusedInstances
+
+/--
+Finds theorems whose bodies were elaborated in the current infotrees and whose (full)
+declaration names satisfy `nameFilter`. Checks their type to see if it contains instance hypotheses
+that (1) are unused in the remainder of the type (2) have types which satisfy `instanceTypeFilter`.
+(Note: `instanceTypeFilter` is non-monadic, and may encounter bound variables in its argument. This
+is a performance optimization. `isAppOrForallOfConstP` may be useful in detecting constant
+applications and types of the form `∀ (...), bar ..` here.)
+
+If any such parameters are found in the type of a theorem `foo`, we create a telescope in which the
+types and free variables of the unused parameters are available as
+`unusedParams : Array Parameter := #[p₁, p₂, ..., pₙ]`, as well as the theorem `thm : ConstantVal`
+and current infotree `t`, and run `log t thm unusedParams`.
+
+A simple pattern is therefore
+```
+fun _ thm unusedParams => do
+  logLint linter.fooLinter (← getRef) m!"\
+    {thm.name.unusedInstancesMsg unusedParams}\n\
+    <extra caption>"
+```
+which logs
+
+> \`{foo}\` has the hypothes[is/es] {p₁}, {p₂}, ..., and {pₙ} which [is/are] not used in the
```

Comment:
```suggestion
> `{foo}` has the hypothes[is/es] {p₁}, {p₂}, ..., and {pₙ} which [is/are] not used in the
```
Why are you "escaping" the backticks?  Is it really necessary here?  If not, there are more escaping slashes later on.

Reply:
Basically, so that what appears when *hovering* over this declaration matches the message generated in the infoview (i.e., when you see the message in the infoview, you see the backtick character, not something code-formatted)...but I'll just make it a code block so it's nice to read in source too. :)